### ISSUE TITLE
suggestion to add new flags for survey shots

### ISSUE THREAD

**Issue by cctrunz** [therion/therion#571]
Hi!

Would it be possible to add a survey shot flag "artificial" for tunnels or other human made underground passages?

We are currently translating some Toporobot survey data to Therion, and we are running into the issue that some passages that were mapped are tunnels (for example for touristic caves). The flags "surface" and "duplicate" are too specific for articial galeries and I feel that its missing something more adequate.

Thanks a lot for looknig into this! and thanks for this great program!

**Comment by robertxa** [comment]
Hi,

I do not know if this is easy to code or not, but I completely agree with cctrunz, it would be very usefull also for me to add new flags for survey shots. I am interesting in "artificial", but also in something like "active meander"/"active galery", "active phreatic galery", "fossil phreatic galery", "active phreatic galery"... 

One suggestion could be that somewhere (in centerline blocks? In a config file (better?)?), the Therion user could define its proper shot flags names with a syntax like:
`def_flags "MyNewFlag" "MNF"` where "MyNewFlag" is the name of the flag use in the command flags, and MNF correspond to the 3-letters string associated to this flag in the sql database.
That means that some cases have to be considered:
- if the new flag definition exist and is used in centerline(s) --> update the database accordingly
- if the new flag definition does not exist and the new flag is used in a centerline: 1) raise a warning during the compilation indicating the presence of an undefined flag, et 2) update the shot database by adding a flag_name by default as "udf" (undefined?).

I will be very interested in that feature for large and long-lived network analysis purposes. I am sure it will strongly extend the possibilities of Therion survey database post-analysis.

Thanks a lot,

Cheers
Xavier

**Comment by clacharm** [comment]
Hello,

I would be interested in preparing a PR for this issue. Before starting the implementation, I would like to confirm the intended scope and syntax.

After reviewing Therion's code and SQL export, it appears that the SQL schema already supports multiple flags for a single survey shot. The table:

```sql
SHOT_FLAG (SHOT_ID, FLAG)
```

already represents a one-to-many relationship, as several rows may reference the same `SHOT_ID`.

The current limitations seem to be located instead in:

- the parser, which accepts only known flags;
- `thdataleg`, which uses a fixed bitmask;
- the SQL exporter, which explicitly handles only `srf`, `dpl`, `apx`, and `spl`.

I would like to propose allowing users to declare custom survey-shot flags, for example:

```therion
shotflag artificial art
shotflag active-gallery acg
```

They could then be used through the existing syntax:

```therion
flags artificial
flags not artificial
```

These declarations would be project-wide, placed at the root level of source files—outside `survey`, `centreline`, and `scrap` blocks—and would have to appear before their first use. Names would be lowercase, and SQL codes would contain exactly three characters.

For an initial version, custom flags would be annotations only:

- they would be attached to the relevant survey shots;
- they would be exported to the existing SQL `SHOT_FLAG` table using their declared three-character codes;
- they would not affect length calculations, loop closure, statistics, rendering, or the handling of native flags;
- native flags would retain their current representation and behaviour.

I also propose that an unknown flag name should produce a compilation error instead of being converted to a generic code such as `udf`, because that would lose the flag's identity and could hide typing mistakes.

An optional table could make the SQL export self-describing:

```sql
SHOT_FLAG_DEF (NAME, FLAG)
```

For example:

```text
artificial     -> art
active-gallery -> acg
```

Before preparing the PR, could you please confirm the following points?

1. Would you prefer this general approach, or only the addition of a built-in `artificial` flag?
2. Does the command name `shotflag <name> <code>` seem appropriate?
3. Would you like a `SHOT_FLAG_DEF` SQL table, or only the codes in the existing `SHOT_FLAG` table?
4. Do you agree that custom flags should initially be annotations without any effect on calculations or rendering?

If this approach is acceptable, I can prepare an atomic PR including tests and the required updates to the Therion Book and `CHANGES`.

For transparency, I plan to use Claude and Codex as development assistants for this PR. I will personally review and understand the proposed changes, run the tests, address review comments, and remain responsible for the submitted code.

Thank you.

**Comment by clacharm** [comment]
A small clarification regarding where `shotflag` declarations could be placed:

The recommended location would remain the project's `thconfig`, either directly:

```therion
shotflag artificial art
shotflag active-gallery acg
```

or through a shared configuration file:

```therion
input project-shotflags.thconfig
```

However, for portability and self-contained datasets, I think the same declaration should also be accepted at the root level of a `.th` source file:

```therion
shotflag artificial art

survey example
  # ...
endsurvey
```

Both locations would feed the same project-wide flag registry.

A declaration in a `.th` file would have to:

- be outside `survey`, `centreline`, and `scrap` blocks;
- appear before its first use;
- use the same validation rules as a declaration from `thconfig`.

Repeating the exact same name-to-code mapping would be harmless, while conflicting mappings would produce a compilation error.

I would still keep these declarations independent of `layout` blocks, since layouts describe map presentation and are not used by `export database`.

Would supporting both `thconfig` and root-level `.th` declarations be acceptable?

**Comment by smudrak** [comment]
Hi Cedric,

first, I am sorry for the delay. For me, I see two possible improvements.

Having new flag "artificial" for surveys of artificial passages in caves excluded from length calculations would be a nice feature.

But having other user defined flags outside of dataset seems to me a little bit complicated. What I would suggest instead is adding custom attributes for survey shots as they can be defined for stations and other objects. Setting for example fossil passage would be:

`shotattr fossil true`

For me, this would fit into existing custom attributes workflows.

Best regards,

S.

**Comment by robertxa** [comment]
Hi Satcho,

Thanks a lot for your answer and suggestions.

I am completely in line with you about a new flag "artificial". 

I haven tough about that, but I also find very interesting your suggestion about defining custom attributes for survey shots instead of shot_flags as proposed here. But I see some constraints and some questions:

- the syntax should probably be something like `shotattr [station1 stationX] fossil true` where `[station1 station2]` is optional and define the shot or the set of shots concerned by the custom shotattr if given (similar to the expand command?); if `[station1 stationX]` is omitted, the custom shotattr apply to the shots written below until the `shotattr fossil false` or `endcenterline` command,
- the custom shot attributes have to be written in the dbf attribues of the shapefiles shots3d and lines2d; each shotattr should be written as a new attribut in the attribut table.
- the custom shot attributes have to be written in the Therion sql export to facilitate further analysis. The idea behind that is to be able to do statistics on various caves with complex geomorphologic and karstification stages histories, and this is my main interest in this development. It could be with the definition of a new field `ATTR: varchar` inside the `SHOT` table. How to write the information in this variable if several `shotattr` are defined for the same shot? Something like `[shotattr1, shotattr2, [...]]`?

Best regards,

Xavier

**Comment by clacharm** [comment]
Hi Stacho, hi Xavier,

Thank you both for the answers. The redirection makes sense to me — I would
rather rework now than defend a design the maintainers do not want.

**Where I stand.** I had a working branch implementing `shotflag <name> <code>`
as described above: declaration object, project-wide registry, propagation to
shots, SQL export, unit tests and a sample test. Nothing is published and no PR
is open, so there is nothing to retract. Reading your two answers, I would drop
the declaration and registry half entirely and keep only the plumbing that
carries a per-shot payload from the centreline parser through to the exporters —
that part is the same whatever the payload turns out to be.

I would then split this into two independent changes.

**1. Native `artificial` flag.** Small and well bounded: one more bit in
`TT_LEGFLAG`, one branch in the length accounting chain in `thdb1d.cxx` (the
`splay / duplicate / surface / else` chain around line 1268), and an `art` code
in `SHOT_FLAG`. Three questions before coding:

- Should artificial legs get their own bucket like surface legs do
  (`stat_slength`), that is counted separately and excluded from `nlength`? And
  should `CENTRELINE` gain an `ARTIFICIAL_LENGTH` column next to
  `SURFACE_LENGTH` and `DUPLICATE_LENGTH`? Without it the SQL export and the log
  would disagree on what the cave is.
- `thmapstat.cxx` sums `stat_length + stat_dlength` for the "surveyed by" and
  "discovered by" tables. Should artificial passages be excluded there too?
- I assume artificial legs still render in maps and in the 3D model exactly like
  normal ones, unlike `surface` legs, and that the flag has no effect on loop
  closure. Is that what you have in mind?

**2. Custom shot attributes.** `shotattr <name> <value>`, mirroring the existing
`attr` workflow. I agree this fits better than a declaration command: no
project-wide registry, no name-to-code table to maintain, and the validation
rules already exist (`th_is_attr_name`, and the leading `_` reserved for system
attributes).

On syntax, Xavier: I would suggest keeping the first version identical in shape
to the existing `flags` command inside `centreline`, that is, state which applies
to the shots that follow until it is changed or the centreline ends:

```therion
centreline
  shotattr fossil true
  1 2 10.0 90 0
  2 3 12.0 95 0
  shotattr fossil false
  ...
```

There is then nothing new to learn, and the optional `[station1 stationX]` range
form you propose can be added later without invalidating any dataset written in
the meantime. Adding it now would give two ways of saying the same thing before
we know which one people actually use.

On the SQL export, one observation that matters for your use case: **today user
attributes do not reach the SQL export at all.** `thexpdb.cxx` creates SURVEY,
CENTRELINE, PERSON, EXPLO, TOPO, STATION, STATION_FLAG, SHOT, SHOT_FLAG, MAPS,
SCRAPS and MAPITEMS, and no attribute table. Station attributes currently go only
to the shapefile DBF, the table export and MetaPost. So the analysis use case
needs new SQL export work whichever design we pick.

For that, I would argue against packing several attributes into a single
`ATTR varchar` column of `SHOT`. A list inside a string has to be parsed back out
in SQL, which defeats the reason for putting it in a database in the first place.
I would rather mirror the shape that already exists for flags:

```sql
create table SHOT_ATTRIBUTE (SHOT_ID integer, NAME varchar, VALUE varchar);
```

One row per attribute, queries stay plain joins, and there is no limit on how
many attributes a shot can carry. If you want the same for stations for
symmetry, `STATION_ATTRIBUTE` would follow the same shape, but that is a separate
change and I would not bundle it here.

For the shapefile export you mention (shots3D, lines2D): that is reachable
through the existing `thattr` machinery, but note that DBF truncates field names
to ten characters (`thattr.cxx:266`), so long attribute names would silently
collide. Worth deciding whether that should be a warning or an error.

**What I would like confirmed before writing code:**

1. Two separate PRs (`artificial` first, `shotattr` after), or one?
2. Is `SHOT_ATTRIBUTE (SHOT_ID, NAME, VALUE)` acceptable, or do you prefer
   another shape?
3. Is `shotattr` the command name you want, or would you rather see it as an
   option of an existing command?
4. For `artificial`: its own length bucket and its own `CENTRELINE` column, or
   simply excluded from the total with no separate reporting?

I will wait for your answers before starting.

Thanks,
Cédric

**Comment by smudrak** [comment]
Hi, for me:

**Two separate PRs (artificial first, shotattr after), or one?**
Definitely two. They are different.

**Is SHOT_ATTRIBUTE (SHOT_ID, NAME, VALUE) acceptable, or do you prefer
another shape?**
OK

**Is shotattr the command name you want, or would you rather see it as an
option of an existing command?**
`shotattr` should be the same as `flags`. So attribute of centerline, not separate command. 

**For artificial: its own length bucket and its own CENTRELINE column, or
simply excluded from the total with no separate reporting?**
I am not sure, if it is needed to be reported. I would simply ignore and create the separete bucket later, when somebody will find usage for it.

S.
